Add files.copy operation

EricDriussi opened the pull request:
<!--
    🎉 Thank you for taking the time to contribute to pyinfra! 🎉

    Please provide a short description of the proposed change and here's a handy checklist of things
    to make PRs quicker to review and merge.

    Note that we will not merge new connectors, but instead welcome PRs that link to thid party
    connector packages.
-->

Adds copy operation suggested in #1121. Any and all feedback is welcome!

- [x] Pull request is based on the default branch (`3.x` at this time)
- [x] Pull request includes tests for any new/updated operations/facts
- [x] Pull request includes documentation for any new/updated operations/facts
- [x] Tests pass (see `scripts/dev-test.sh`)
- [x] Type checking & code style passes (see `scripts/dev-lint.sh`)

EricDriussi committed:
operations/files.copy: basic copy functionality

EricDriussi committed:
operations/files.copy: copy dirs

EricDriussi committed:
operations/files.copy: docs

DonDebonair commented on src/pyinfra/operations/files.py:
```diff
@@ -1312,6 +1312,35 @@ def move(src: str, dest: str, overwrite=False):
+    dest_file_path = os.path.join(dest, os.path.basename(src))
+    dest_file_exists = host.get_fact(File, dest_file_path)
+    if dest_file_exists and not overwrite:
+        raise OperationError(f"dest {dest_file_path} already exists and `overwrite` is unset")
```
Why should this raise an `OperationError`. For `copy()` to be idempotent, the destination file existing, should just be a _noop_ right? Alternatively, you could compare hashes to see if the src file has changed.

EricDriussi replied:
Indeed a noop would be a better fit here, although I still feel that an error should be raised in case `dest_file_path` already exists **and** is different from `src`. So maybe something like this?

```python
    if dest_file_exists and not overwrite:
        if _file_equal(src, dest_file_path):
            host.noop(f"{dest_file_path} already exists")
        else:
            raise OperationError(f"{dest_file_path} already exists and is different than src")
```

Or maybe there's a better way to handle that case?

Also, in the snippet I use the `_file_equal` [function](https://github.com/pyinfra-dev/pyinfra/blob/6bc88997705a34976d8b2c47f1e2d94e0f187c4c/pyinfra/operations/files.py#L742C1-L756C1), but that seems to be intended for local-remote hash comparisons.
Are you aware of any functions that might better fit this use case? Should I just create a new internal function based on this one for remote-remote hash comparison (what could be a nice name for it?)?

DonDebonair replied:
I agree with you that the noop should only happen when the existing file at destination is the same.

I don't know if the `_file_equal` function can be used for remote-remote hash comparisons. @Fizzadar do you know?

Fizzadar replied:
Indeed `_file_equal` is for local <> remote, a new function seems best here, basically fact x2 and compare.

EricDriussi committed:
operations/files.copy: noop if file already exists

EricDriussi committed:
operations/files.copy: compare files locally

EricDriussi committed:
operations/files.copy: skip win in tests

EricDriussi commented:
Hey! Sorry it took me forever to get back to this...
I implemented the [changes mentioned here](https://github.com/pyinfra-dev/pyinfra/pull/1446#discussion_r2328793654), added/fixed tests and skipped the Windows CI (simply due to me not having a Windows machine to ensure the tests work before pushing).

Also, in the new [`_remote_file_equals` function](https://github.com/pyinfra-dev/pyinfra/pull/1446/files#diff-86e5946f52cfcf78fe5a111af41e331fe33ffc0aba21766b254b4eb1a8e1112aR754), I followed the same fallback logic I saw in `_file_equals` regarding the checksum comparison.
I'm not quite sure if this is correct so please let me know if there's anything that needs to be updated before merging.

Fizzadar reviewed (approved):
@EricDriussi no worries re:times, thank you for finishing this up! Looks good to me will leave for Daan to have a double check too.
